rocAL/include/readers/image/numpy_data_reader.h

```diff
@@ -0,0 +1,154 @@
+    std::string id() override { return _last_id; };
+
+    //! Returns the name of the latest file_path opened
+    const std::string file_path() override { return _last_file_path; }
+
+    unsigned count_items() override;
+
+    ~NumpyDataReader() override;
+
+    int close() override;
+
+    NumpyDataReader();
+
+    //! Returns the number of images in the last batch
+    size_t last_batch_padded_size() override;
+
+    std::string get_root_folder_path();  // Returns the root folder path
+
+    std::vector<std::string> get_file_paths_from_meta_data_reader();  // Returns the relative file path from the meta-data reader
+
+   private:
+    //! opens the folder containnig the images
+    Reader::Status open_folder();
+    Reader::Status subfolder_reading();
+    std::string _folder_path;
+    DIR* _src_dir;
+    DIR* _sub_dir;
+    struct dirent* _entity;
+    std::vector<std::string> _file_names;
+    unsigned _curr_file_idx;
+    FILE* _current_fPtr;
+    unsigned _current_file_size;
+    unsigned _shard_start_idx;
+    NumpyHeaderData _curr_file_header;
+    std::string _last_id;
+    std::string _last_file_name, _last_file_path, _absolute_file_path;
+    size_t _shard_id = 0;
+    size_t _shard_count = 1;  // equivalent of batch size
+    signed _shard_size = -1;
+    //!< _batch_count Defines the quantum count of the images to be read. It's usually equal to the user's batch size.
+    /// The loader will repeat images if necessary to be able to have images available in multiples of the load_batch_count,
+    /// for instance if there are 10 images in the dataset and _batch_count is 3, the loader repeats 2 images as if there are 12 images available.
+    size_t _batch_count = 1;
+    size_t _file_id = 0;
+    size_t _padded_samples = 0;
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    //!< _file_count_all_shards total_number of files in to figure out the max_batch_size (usually needed for distributed training).
+    size_t _file_count_all_shards;
+    std::mutex _cache_mutex_;
+    std::map<std::string, NumpyHeaderData> _header_cache_;
+    const RocalTensorDataType TypeFromNumpyStr(const std::string& format);
+    inline void SkipSpaces(const char*& ptr);
+    void ParseHeaderContents(NumpyHeaderData& target, const std::string& header);
+    template <size_t N>
+    void Skip(const char*& ptr, const char (&what)[N]);
+    template <size_t N>
+    bool TrySkip(const char*& ptr, const char (&what)[N]);
+    template <size_t N>
+    void SkipFieldName(const char*& ptr, const char (&name)[N]);
+    template <typename T = int64_t>
+    T ParseInteger(const char*& ptr);
+    std::string ParseStringValue(const char*& input, char delim_start = '\'', char delim_end = '\'');
+    void ParseHeader(NumpyHeaderData& parsed_header, std::string file_path);
+    template <typename T>
+    size_t ParseNumpyData(T* buf, std::vector<unsigned> strides, std::vector<unsigned> shapes, unsigned dim = 0);
```

Comment:
All above numpy functions mus necessarily be part of the class?
Can they be independent functions  in Cpp file is that feasible or can it be the right approach please check